docs/best-practice-consensus-pipeline.md

```diff
@@ -0,0 +1,217 @@
+      --dont-store-original-umis
+```
+
+Neither option above requires or performs any sorting.
+
+### Step 1.2: uBam -> Mapped BAM
+
+This step uses a combination of `samtools` (to convert the uBam back to FASTQ), `bwa mem` to do the actual alignment and `fgbio ZipperBams` to carry metaadata and UMIs forward from the unmapped BAM.  No sorting is performed in this step.
+
+```  
+# Align the data with bwa and recover headers and tags from the unmapped BAM
+samtools fastq s1.unmapped.bam \
+  | bwa mem -t 16 -p -K 150000000 -Y ref.fa \
+  | fgbio -Xmx4g --compression 1 --async-io ZipperBams \
+      --unmapped-bam s1.unmapped.bam \
+      --ref ref.fa \
+      --output s1.mapped.bam 
+```
+
+Some important `bwa mem` options used here are:
+
+- `-t 16` to set the number of threads to use to 16; this can be set to any number between 1 and the number of CPU cores available
+- `-K 150000000` to tell `bwa` to process reads/bases in fixed size chunks regardless of the number of threads; this makes `bwa` deterministic and reproducible with different thread counts
+- `-Y` to use soft-clipping (instead of hard clipping) in supplemntal alignments
+
+Under no circumstances should `-M` be used to `mark shorter split hits as secondary` as this will cause problems with downstream tools.
+
+
+### Step 1.3: Mapped BAM -> Grouped BAM
+
+This step identifies reads or read pairs that originate from the same source molecule based on genomic positions and UMI.  For _most_ data this is the recommended way to run grouping:
+
+```  
+# Group the reads by position and UMI
+fgbio -Xmx8g --compression 1 GroupReadsByUmi \
+  --input s1.mapped.bam \
+  --strategy Adjacency \
+  --edits 1 \
+  --output s1.grouped.bam \
+  --family-size-histogram s1.tag-family-sizes.txt
+```
+
+However:
+
+1. If UMIs are significantly longer (e.g. 20bp) or have more errors, `--edits` can be increased
+2. For multiplex PCR and similar data where reads' genomic positions are fixed by the primers it is recommended to use `--strategy Identity` to reduce runtime at the expense of lower accuracy
+
+
+## Phase 2(a): GroupedBam -> Filtered Consensus, R&D Version
+
+
+```mermaid
+graph TD;
+A["fgbio CallMolecularConsensusReads"]-->B["samtools fastq | bwa mem | fgbio ZipperBam"];
+B-->C["fgbio FilterConsensusReads | samtools sort"];
+```
+
+### Step 2(a).1: GroupedBam -> Consensus uBam
+
+This step generates unmapped consensus reads from the grouped reads.  Depending on your data you may choose to raise or lower the `--min-input-base-quality` to accept or reject more bases from the grouped raw reads.  In an R&D setting it is frequently useful to see _all_ consensus reads, but there is a performance trade-off.  Frequently a very large fraction of molecules will have only a single read[-pair]; if consensus reads with depth 1 will be discarded later setting `--min-reads 2` can cause this step to run significantly faster.
+
+```bash
+fgbio -Xmx4g --compression 0 CallMolecularConsensusReads \
```

Comment:
async-io please

Reply:
Added a section on asyncio.  I think there is near-0 benefit when using compression 0 so leaving it out here.  Especially with `--threads`, most of the work is in a threadpool, leaving the main thread idle to write the BAM.